### PR TITLE
[fix](nereids) fix some arrgregate bugs in Nereids

### PR DESCRIPTION
**Pull request by starocean999**
# Proposed changes

Issue Number: close #xxx

## Problem summary
1. the agg function without distinct keyword should be a "merge" funcion in threePhaseAggregateWithDistinct
2. use aggregateParam.aggMode.consumeAggregateBuffer instead of aggregateParam.aggPhase.isGlobal() to indicate if a agg function is a "merge" function
3. add an AvgDistinctToSumDivCount rule to support avg(distinct xxx) in some case
4. AggregateExpression's nullable method should call inner function's nullable method.
5. add a bind slot rule to bind pattern "logicalSort(logicalHaving(logicalProject()))"
6. don't remove project node in PhysicalPlanTranslator
7. add a cast to bigint expr when count( distinct datelike type )
8. fallback to old optimizer if bitmap runtime filter is enabled.
 

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.94 seconds
 load time: 637 seconds
 storage size: 17123290494 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221223145245_clickbench_pr_68099.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PushdownExpressionsInHashCondition.java [hunk 1]
```diff
@@ -79,7 +79,9 @@ public Rule build() {
                         exprsOfHashConjuncts.get(0).add(conjunct.child(0));
                         exprsOfHashConjuncts.get(1).add(conjunct.child(1));
                         conjunct.children().forEach(expr ->
-                                exprMap.put(expr, new Alias(expr, "expr_" + expr.toSql())));
+                                exprMap.put(expr,
+                                        expr instanceof Slot ? new Alias(((Slot) expr).getExprId(), expr, expr.toSql())
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PushdownExpressionsInHashCondition.java:83]: make this logical into Alias constructor

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PushdownExpressionsInHashCondition.java:83]: how about do not create this useless project?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AvgDistinctToSumDivCount.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * AvgDistinctToSumDivCount.
+ *
+ * change avg( distinct a ) into sum( distinct a ) / count( distinct a ) if there are more than 1 distinct arguments
+ */
+public class AvgDistinctToSumDivCount extends OneAnalysisRuleFactory {
+    @Override
+    public Rule build() {
+        return RuleType.AVG_DISTINCT_TO_SUM_DIV_COUNT.build(
+                logicalAggregate().then(agg -> {
+                    if (agg.getDistinctArguments().size() > 1) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AvgDistinctToSumDivCount.java:48]: nit: could use when:
```suggestion
                logicalAggregate().when(agg -> agg.getDistinctArguments().size() > 1).then(agg -> {
                    ...
```

Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AvgDistinctToSumDivCount.java:48]: nit: could use when:
```suggestion
                logicalAggregate().when(agg -> agg.getDistinctArguments().size() > 1).then(agg -> {
                    ...
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateStrategies.java [hunk 3]
```diff
@@ -781,7 +781,7 @@ private List<PhysicalHashAggregate<? extends Plan>> twoPhaseAggregateWithDistinc
                 false, Optional.empty(), logicalAgg.getLogicalProperties(),
                 requireGather, logicalAgg.child());
 
-        AggregateParam inputToResultParam = new AggregateParam(AggPhase.GLOBAL, AggMode.INPUT_TO_RESULT);
+        AggregateParam inputToResultParam = new AggregateParam(AggPhase.GLOBAL, AggMode.BUFFER_TO_RESULT);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateStrategies.java:784]:
```suggestion
        AggregateParam bufferToResultParam = new AggregateParam(AggPhase.GLOBAL, AggMode.BUFFER_TO_RESULT);
```

Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateStrategies.java:784]:
```suggestion
        AggregateParam bufferToResultParam = new AggregateParam(AggPhase.GLOBAL, AggMode.BUFFER_TO_RESULT);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java [hunk 4]
```diff
@@ -51,15 +56,19 @@ public Rule build() {
                 predicate = new EqualTo(((InSubquery) apply.getSubqueryExpr()).getCompareExpr(),
                         apply.right().getOutput().get(0));
             }
-
+            List<Expression> conjuncts = ExpressionUtils.extractConjunction(predicate);
+            if (conjuncts.stream().anyMatch(expression -> expression.children().stream()
+                    .anyMatch(expr -> expr.getDataType() == BitmapType.INSTANCE))) {
+                throw new NereidsException(new AnalysisException("nereids don't support bitmap filter"));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java:62]: add a TODO on it, we will fix it later.
```suggestion
                throw new AnalysisException("nereids don't support bitmap filter");
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/MultiDistinctCount.java [hunk 5]
```diff
@@ -18,28 +18,39 @@
...
 /** MultiDistinctCount */
 public class MultiDistinctCount extends AggregateFunction
         implements AlwaysNotNullable, ExplicitlyCastableSignature {
     public MultiDistinctCount(Expression arg0, Expression... varArgs) {
-        super("multi_distinct_count", true, ExpressionUtils.mergeArguments(arg0, varArgs));
+        super("multi_distinct_count", true, ExpressionUtils.mergeArguments(arg0, varArgs).stream()
+                .map(arg -> arg.getDataType() instanceof DateLikeType ? new Cast(arg, BigIntType.INSTANCE)
+                        : arg)
+                .collect(
+                        Collectors.toList()));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/MultiDistinctCount.java:45]: i think we should change getSignatures, return BigIntType on DateLikeType args.
```suggestion
        super("multi_distinct_count", true, ExpressionUtils.mergeArguments(arg0, varArgs).stream()
                .map(arg -> arg.getDataType() instanceof DateLikeType
                        ? new Cast(arg, BigIntType.INSTANCE) : arg)
                .collect(Collectors.toList()));
```

**Comment by 924060929** [comment]
Hi, I have some question.
1. why we must use AvgDistinctToSumDivCount to process the avg(distinct)
2. a digression, I think you should merge ProjectWithDistinctToAggregate to ProjectToGlobalAggregate

**Comment by morrySnow** [comment]
> 2. why shouldn't eliminate project in the translator?

i can answer the 2nd question.
currently, we use two map to maintain ExprId and SlotId's relationship. However, in some cases, this cannot work fine when we eliminate project in translator. Instead, we need to maintain a 1 to n relationship. And in some other cases, we find more than one bugs in this feature's code.
So, we just remove this feature temporarily and add it back later~

**Review by morrySnow — APPROVED** [review]

**Pull request merged**